torch/csrc/jit/passes/constant_propagation.cpp

```diff
@@ -193,11 +250,29 @@ struct ConstantPropagator {
     }
   }
 
-  void ConstantPropagation(Node* n) {
-    bool constant_inputs =
-        std::all_of(n->inputs().begin(), n->inputs().end(), [&](Value* v) {
+  // An Op has runnable inputs if:
+  // - All inputs are constants.
+  // - It is an op that forwards tuples, and all inputs are constants
+  // or tuples that we know the ivalue for. We can't use known tuple ivalues
+  // for non-forwarding ops because that Tuple could contain an ivalue that is
+  // not allowed as a constant, for instance, a Tensor with a gradient.
+  bool runnableInputs(Node* n) {
+    if (std::all_of(n->inputs().begin(), n->inputs().end(), [&](Value* v) {
           return v->node()->kind() == prim::Constant;
-        });
+        })) {
+      return true;
+    }
+    if (tuple_ops.count(n->kind())) {
+      return (
+          std::all_of(n->inputs().begin(), n->inputs().end(), [&](Value* v) {
+            return v->node()->kind() == prim::Constant || tuples.count(v);
```

Comment:
Won't `tuples.count(v)` return true for any tuple since contains all of them regardless of whether they are constant or not

Reply:
No, it will only return true for tuples which we've computed the ivalue for.